core/src/main/java/hudson/model/Descriptor.java

```diff
@@ -918,17 +918,29 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
                 // Descriptors, so we prefer 'kind' if it's present.
                 String kind = jo.optString("kind", null);
                 if (kind != null) {
+                    // Only applies when Descriptor.getId is overridden.
+                    // Note that kind is only supported here,
+                    // *not* inside the StaplerRequest.bindJSON which is normally called by newInstance
+                    // (since Descriptor.newInstance is not itself available to Stapler).
+                    // If you merely override getId for some reason, but use @DataBoundConstructor on your Describable,
+                    // there is no problem; but you can only rely on newInstance being called at top level.
                     d = findById(descriptors, kind);
                 }
                 if (d == null) {
-                  kind = jo.getString("$class");
-                  d = findByDescribableClassName(descriptors, kind);
-                  if (d == null) d = findByClassName(descriptors, kind);
+                  kind = jo.optString("$class");
+                  if (kind != null) { // else we will fall through to the warning
+                      // This is the normal case.
+                      d = findByDescribableClassName(descriptors, kind);
+                      if (d == null) {
+                          // Deprecated system where stapler-class was the Descriptor class name (rather than Describable class name).
+                          d = findByClassName(descriptors, kind);
+                      }
+                  }
                 }
                 if (d != null) {
                     items.add(d.newInstance(req, jo));
                 } else {
-                    LOGGER.warning("Received unexpected formData for descriptor " + kind);
+                    LOGGER.log(Level.WARNING, "Received unexpected form data element: {0}", jo);
```

Comment:
Might this warning be a bit misleading if the previous code fails to find the descriptor (`d`) from the `$class` value? Seems to me that a warning like ` .... "Failed to find descriptor for $class {0}", kind` would be more helpful (than "unexpected form data element").